---
title: 20181008 TIL
date: 2018-10-08 19:41:49
tags: swift, mutating, throw, do-try-catch
---

### mutating
- 인스턴스에 부여된 struct의 property를 변형할 수 있는 keyword

```swift
struct meal {
    private(set) var menu: String = "김치찌개"

    mutating func goToSubway() {
        menu = "이탈리안BMT"
    }
}

let lunch = meal()
print(lunch.menu) //"김치찌개"

lunch.goToSubway()
print(lunch.menu) //"이탈리안BMT"
```

### throw
- 발생 가능한 error에 대해, handling을 위해 swift에서 미리 준비해둔 기능들. 특정 조건 및 환경에서 에러를 던져주고, 나중에 해당 에러를 캐치하여 처리할 수 있다.
- 에러를 던지는 메소드는 try로만 calling할 수 있다. 이후에 적절한 처리가 온다는 가정하에…

```swift
enum TimeError : Error {
    case invalidTime
    case nightTime
}

func stretchBody(recess: Time) throws -> Void {      // 에러를 던질수 있는 메서드에서만 throw를 쓸 수 있다. 표현은 argument 뒤에 throws를 써준다.

    guard let restTime = recess else {
        throw TimeError.invalidTime //에러 타입의 에러를 던진다.
    }
}
```

### do - try - catch
- throw로 던져지는 에러들을 받아서 처리하는 구문. 스위치와 비슷하게 생겼다.

```swift
var noon: Time? = 12
noon = nil

do {
    try stretchBody(recess: noon)      // 실행하고자하는, 오류가 발생할 수 있는 코드를 입력합니다. 여러줄 가능.
} catch TimeError.invaliedTime {
    print("유효하지 않는 시간입니다.")
} catch TimeError.nightTime {
    print("자고있는 시간입니다.")
} catch {
    print("예상치 못한 에러입니다 : \(error)")     // catch 뒤에 에러패턴을 입력하지 않으면 error 라는 지역상수가 생성됩니다.
}
```

***

문법을 아는것과, 구조를 설계하는 것에는 갭이 큰 것 같다… 에러를 던져야 하는 부분이 산재해 있어서 두 캐치도 여기저기 쓰게 되는 것이 마음에 들지 않아서, 공부 했는데도 이번에는 써먹을 수 없었다. 미리 고려하며 짜는 것이 중요할 듯
